Parity integration testing

dylanjw opened the pull request:
### What was wrong?

There are no parity integration tests: see https://github.com/ethereum/web3.py/issues/381

### How was it fixed?

Added:
a parity install script
parity fixtures generation script
parity pytests

TODO:
- [x] configure travis CI to run tests against variety of parity versions

#### Cute Animal Picture

![Cute animal picture](https://i.pinimg.com/736x/a9/eb/71/a9eb71a5266c3ab1e9b9affd737c065e--funny-animals-timeline-photos.jpg)

pipermerriam commented on tests/integration/generate_fixtures/common.py:
```diff
@@ -0,0 +1,240 @@
+    Make sure that a path exists
+    """
+    if not os.path.exists(dir_path):
+        os.makedirs(dir_path)
```
FYI, this entire utility is replacable by just using `os.makedirs(the_path, exist_ok=True)`

pipermerriam commented on tests/integration/generate_fixtures/common.py:
```diff
@@ -0,0 +1,240 @@
+    return str(port)
+
+
+def get_geth_binary():
```
does this belong in `go_ethereum.py`?

pipermerriam commented on tests/integration/generate_fixtures/common.py:
```diff
@@ -0,0 +1,240 @@
+
+
+@contextlib.contextmanager
+def get_geth_process(geth_binary,
```
Same her, `go_ethereum.py` instead of `common.py`?

dylanjw replied:
I put those in common.py because I use them in the parity script to create the cluster.

pipermerriam replied:
Took me a minute to figure out what was going on.  That makes sense now.  +1 for clever way to de-duplicate the logic for initializing the chain.

dylanjw commented:
The tests are running in travis now, but failing due to the missing ipc file, which probably means parity closed unexpectedly.  Unfortunately im not seeing any stdout from parity in the travis log.  Next, I will work on getting the parity console output to print to the travis log.

pipermerriam commented:
> The tests are running in travis now, but failing due to the missing ipc file

I admittedly didn't look at the test, but... you may have to put in some `time.sleep()` statements in to wait for the IPC file to exist as parity initializes.  If you're already doing this then disregard.

dylanjw commented:
I got this working on a local docker instance of the travis python container.  It seems that the contents of the parity base directory is not sufficient to transfer the private chain data between environments.  I added import/export commands to the fixture creation and test scripts, which solved the first issue.

Now, the build times out while downloading parity due the absence of output for 10 minutes.  I have added a progress bar to install_parity.py to prevent the travis timeout.

carver commented on .travis.yml:
```diff
@@ -58,6 +64,7 @@ before_install:
 install:
   - travis_retry pip install pip setuptools --upgrade
   - travis_retry pip install tox
+  - travis_retry pip install eth_utils tqdm requests
```
Since these are only necessary during parity install, we could skip it in all the tests by inserting the `pip install` as a line with its own: `if [ -n "$PARITY_VERSION" ];` just before the `install_parity` command.

carver commented:
It will make me so happy to have these in web3!

pipermerriam commented:
Next up: Trinity ✨

carver commented on .travis.yml:
```diff
@@ -53,11 +60,13 @@ cache:
 install:
   - travis_retry pip install pip setuptools --upgrade
   - travis_retry pip install tox
+  - if [ -n "$PARITY_VERSION" ]; travis_retry pip install eth_utils tqdm requests
```
looks to be missing ending: `; fi`

carver commented on .circleci/config.yml:
```diff
@@ -15,25 +15,33 @@ common: &common
-            sudo apt-get install -y build-essential;
-            python -m geth.install $GETH_VERSION;
-            sudo ln -s /home/circleci/.py-geth/geth-v1.7.2/bin/geth /usr/local/bin/geth
+          if [ -n "$GETH_VERSION" ]; then
```
I seem to remember there being a problem with doing this, like it messed up the cache. Maybe it would work if you add the GETH_VERSION environment variable to the cache key. Note that because the cache was already built successfully, changes here are dangerous: we might break things and not realize it.

carver commented on .circleci/config.yml:
```diff
@@ -15,25 +15,33 @@ common: &common
+    - run:
+        name: install parity if needed
+        command: |
+          if [ -n "$PARITY_VERSION" ]; then
```
Same thing here, might want to add PARITY_VERSION to the cache key.

dylanjw commented:
Im going to move any circle CI config changes from this PR into a new one where they can be tested.  It seems like circle ci is only picking up new pull requests.
